tokio-executor/src/task/harness.rs

```diff
@@ -0,0 +1,592 @@
+        mut self,
+        executor: *const S,
+        join_interest: bool,
+        out: Track<super::Result<T::Output>>,
+    ) {
+        if join_interest {
+            // Store the output. The future has already been dropped
+            unsafe {
+                self.store_out(out);
+            }
+        }
+
+        let bound_executor = unsafe { self.header().executor.with(|ptr| *ptr) };
+
+        // Handle releasing the task. First, check if the current
+        // executor is the one that is bound to the task:
+        if executor == bound_executor {
+            unsafe {
+                // perform a local release
+                let task = ManuallyDrop::new(Task::from_raw(self.header));
+                (*bound_executor).release_local(&task);
+
+                if self.transition_to_released(join_interest).is_final_ref() {
+                    self.dealloc();
+                }
+            }
+        } else {
+            let res = self.transition_to_complete(join_interest);
+            assert!(!res.is_final_ref());
+
+            if res.has_join_waker() {
+                // The release step happens later once the task has migrated back to
+                // the worker that owns it. At that point, the releaser **may** also
+                // be responsible for dropping. This fact must be tracked until
+                // the release step happens.
+                self.set_might_drop_join_waker_on_release();
+            }
+
+            unsafe {
+                let task = Task::from_raw(self.header);
+                (*bound_executor).release(task);
+            }
+        }
+    }
+
+    /// Return `true` if the task structure should be deallocated
+    fn transition_to_complete(&mut self, join_interest: bool) -> Snapshot {
+        let res = self.header().state.transition_to_complete();
+
+        self.notify_join_handle(join_interest, res);
+
+        // Transition to complete last to ensure freeing does
+        // not happen until the above work is done.
+        res
+    }
+
+    /// Return `true` if the task structure should be deallocated
+    fn transition_to_released(&mut self, join_interest: bool) -> Snapshot {
+        if join_interest {
+            let res1 = self.transition_to_complete(join_interest);
+
+            // At this point, the join waker may not be changed. Once we perform
+            // `release_task` we may no longer read from the struct but we
+            // **may** be responsible for dropping the waker. We do an
+            // optimistic read here.
+            let (join_waker, check) = unsafe { self.read_join_waker() };
+
+            let res2 = self.header().state.release_task();
+
+            if res1.has_join_waker() && !res2.is_join_interested() {
+                debug_assert!(res2.has_join_waker());
+
+                // Its our responsibility to drop the waker
+                check.check();
+                unsafe {
+                    let _ = join_waker.assume_init();
```

Comment:
Nit/TIOLI:
```suggestion
                    drop(join_waker.assume_init());
```
might express intent better?